standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaStoreChecker.java

```diff
@@ -311,48 +315,96 @@ void checkTable(Table table, PartitionIterable parts, byte[] filterExp, CheckRes
+        if (partPath == null) {
+          continue;
+        }
+        futures.add(pool.submit(new Callable() {
+          @Override
+          public Object call() throws Exception {
+            Path tempPartPath = partPath;
+            FileSystem tempFs = tempPartPath.getFileSystem(conf);
+            CheckResult.PartitionResult prFromMetastore = new CheckResult.PartitionResult();
+            String partName = getPartitionName(table, partition);
+            prFromMetastore.setPartitionName(partName);
+            prFromMetastore.setTableName(partition.getTableName());
+
+            synchronized (this) {
+              if (!tempFs.exists(tempPartPath)) {
+                result.getPartitionsNotOnFs().add(prFromMetastore);
+              } else {
+                result.getCorrectPartitions().add(prFromMetastore);
+              }
+            }
 
-      for (int i = 0; i < getPartitionSpec(table, partition).size(); i++) {
-        Path qualifiedPath = partPath.makeQualified(fs);
-        partPaths.add(qualifiedPath);
-        partPath = partPath.getParent();
+            if (partitionExpirySeconds > 0) {
+              long currentEpochSecs = Instant.now().getEpochSecond();
+              long createdTime = partition.getCreateTime();
+              long partitionAgeSeconds = currentEpochSecs - createdTime;
+              if (partitionAgeSeconds > partitionExpirySeconds) {
+                CheckResult.PartitionResult pr = new CheckResult.PartitionResult();
+                pr.setPartitionName(getPartitionName(table, partition));
+                pr.setTableName(partition.getTableName());
+                synchronized (result) {
+                  result.getExpiredPartitions().add(pr);
+                }
+                if (LOG.isDebugEnabled()) {
+                  LOG.debug("{}.{}.{}.{} expired. createdAt: {} current: {} age: {}s expiry: {}s", partition.getCatName(),
+                      partition.getDbName(), partition.getTableName(), pr.getPartitionName(), createdTime, currentEpochSecs,
+                      partitionAgeSeconds, partitionExpirySeconds);
+                }
+              }
+            }
+
+            synchronized (this) {
+              for (int i = 0; i < getPartitionSpec(table, partition).size(); i++) {
+                final Path qualifiedPath = tempPartPath.makeQualified(tempFs);
+                partPaths.add(qualifiedPath);
+                tempPartPath = tempPartPath.getParent();
+              }
+            }
+            return processedPartitions.add(partName);
+          }
+        }));
+      }
+      while (!futures.isEmpty()) {
+        futures.poll().get();
+      }
+    } catch (ExecutionException | InterruptedException e) {
+      LOG.error("Exception occurred while processing partitions "+e);
+    }
+    finally {
+      LOG.debug("Partition exists check complete for table "+table.getCatName()+"."+table.getDbName()+"."+table.getTableName()+" partitions "+processedPartitions);
```

Comment:
`LOG.debug("Partition exists check complete for table {}.{}.{} partitions {}", table.getCatName(), table.getDbName(), table.getTableName(), processedPartitions);`